File: Data Structures & Algorithms/binary-search/submission-2.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = -1
        right = len(nums)
        while left + 1 < right:
            mid = right + (left - right) // 2
            #严格维护
            #nums[left] < target
            #nums[right] >= target
            if nums[mid] >= target:
                right = mid
            else:
                left = mid
        return right if right != len(nums) and nums[right] == target else -1


    

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int ans = 0;
    public int goodNodes(TreeNode root) {
        dfs(root, root.val);
        return ans;
    }
    private void dfs(TreeNode root, int maxVal) {
        if(root == null) return;
        if(root.val >= maxVal) ans++;
        maxVal = Math.max(maxVal, root.val);
        dfs(root.left, maxVal);
        dfs(root.right, maxVal);
    }
}
